### PR TITLE
Topic recognition

### PR DESCRIPTION
**Pull request by rljinuq**
Name: Lingjie Ruan
Student ID: 47640960

Those files contain my model and report for the task, Semi-supervised Multi-class Node Classification via GCN.

Specifically, it adds the folder s4764096_GCN, with the following files alongside README.md:

dataset.py: contains functions for loading and preparing the graph data for training and testing.
modules.py: defines the GCN model architecture and its supporting layers.
train.py: contains functions for training the GCN model, tracking loss, accuracy, and plotting results.
predict.py: loads the trained model and visualizes the learned embeddings using UMAP.
The folder also contains the following images necessary for the README:

p1.png: UMAP visualization of the node embeddings before training.
p2.png: UMAP visualization of the node embeddings after training.
p3.png: UMAP visualization after 100 epochs of training.
p4.png: Plot showing the Training Loss and Test Accuracy over 100 epochs.

### CODE REVIEW

**Comment by gayanku** [comment]
**This is an initial inspection, no action is required at this point**
GNN FB Page-Page Network dataset ---- Normal Difficulty
| Category	| | Marks | Comments |	
| :---  | :---: | :---: | :---  |		
| Algorithm solves the problem 		| 5 | 4.5 | No early stop|
| Implementation functions as intended	| 3 | 3 | 4 layer GCN, dglnn.GraphConv |
| Good design 					| 1 | 1 | Modular, Reusable |
| Commenting 					| 1 | 1 | Meaningful docstrings and comments |
| Algorithm above Normal Difficulty 	| 5 | 5 | |
| Algorithm is Hard difficulty 			| 5 | 0 | Normal Difficulty |
| Section IV : Max mark 15 from 20		|   | **14.5**   | |

**TSNE / UMAP:** 
Good.
- Discussion: Good, with a bit more depth useful.

**Suggestions:**
Would be useful to have test/train loss and test/train acc on separately to identify overfitting etc.
Code can have some early stop, ex: train loop should save best model by keeping track of current best accuracy.
Could add a hyper parameter search or discuss how the values were obtained.

**Comment by gayanku** [comment]
"## Marking
<table><tbody>
  <tr><td colspan='3'><b>Good/OK/Fair Practice (Design/Commenting, TF/Torch Usage)</b></td> </tr>
	<tr><td></td><td>Good design and implementation. </td><td></td></tr>
	<tr><td></td><td>Spacing and comments. </td><td></td></tr>
	<tr><td></td><td>No Header blocks. </td><td>-1</td></tr>
	
  <tr><td colspan='3'><b>Recognition Problem</b></td></tr>
	<tr><td></td><td>OK solution to problem. No early stop.</td><td>-1</td></tr>
	<tr><td></td><td>Driver Script present. </td><td></td></tr>
	<tr><td></td><td>File structure present. </td><td></td></tr>
	<tr><td></td><td>Good Usage & Demo & Visualisation & Data usage. </td><td></td></tr>
	<tr><td></td><td>Module present. </td><td></td></tr>
	<tr><td></td><td>Commenting missing. </td><td>-1</td></tr>
	<tr><td></td><td>No Data leakage found. </td><td></td></tr>
	<tr><td></td><td>Difficulty : Normal. GNN Task.</td><td>-5</td></tr>
	
  <tr><td colspan='3'><b>Commit Log</b></td></tr>
	<tr><td></td><td>Some/Adequate Meaningful commit messages. Basic details only. 10commits, 4 days.</td><td>-1</td></tr>
	<tr><td></td><td>Some/Adequate Progressive commits. Minimal.</td><td>-1</td></tr>
	
  <tr><td colspan='3'><b>Documentation</b></td></tr>
	<tr><td></td><td>Readme :Acceptable. </td><td>-1</td></tr>
	<tr><td></td><td>Model/technical explanation :Acceptable. Model infor missing</td><td>-1</td></tr>
	<tr><td></td><td>Description and Comments :Good. </td><td></td></tr>
	<tr><td></td><td>Markdown used and PDF submitted. </td><td></td></tr>
	
  <tr><td colspan='3'><b>Pull Request</b></td></tr>
	<tr><td></td><td>Successful Pull Request (Working Algorithm Delivered on Time in Correct Branch). </td><td></td></tr>
	<tr><td></td><td>No Feedback required. </td><td></td></tr>
	<tr><td></td><td>Request Description is adequate. Add some info about the specific implementation here.</td><td>-1</td></tr>
  <tr><td colspan='2'><b>TOTAL</b></td><td>-13</td></tr>
</tbody></table>

Marked as per the due date and changes after which aren't necessarily allowed to contribute to grade for fairness.
_Subject to approval from Shakes_
